fmn/rules/services/distgit.py

```diff
@@ -82,22 +93,44 @@ def invalidate_on_message(self, message: "Message"):
             ]
             owners = list(chain(*owner_accesses))
             if message.body["new_group"] in owners:
-                self._on_owner_changed(
+                await self._on_owner_changed(
                     message.body["project"]["namespace"],
                     message.body["project"]["name"],
                     message.body["new_group"],
                     "group",
                 )
         elif message.topic.endswith("pagure.project.group.removed"):
             if message.body["access"] in self.GROUP_OWNER_LEVELS:
-                self._on_owner_changed(
+                await self._on_owner_changed(
                     message.body["project"]["namespace"],
                     message.body["project"]["name"],
                     message.body["new_group"],
                     "group",
                 )
 
-    def _on_owner_changed(self, namespace, project_name, name, user_or_group):
-        self.get_owners.refresh(self, namespace, project_name, user_or_group)
-        self.get_owned.refresh(self, namespace, name, user_or_group)
-        cache.invalidate_tracked()
+    async def _on_owner_changed(self, namespace, project_name, name, user_or_group):
+        # Don't use the cache.invalidate() decorator as it will invalidate the cache *after* the
+        # function has been run, and we want to refresh it here.
+        await cache.invalidate_func(
+            self.get_owners,
+            {
+                "artifact_type": namespace,
+                "artifact_name": project_name,
+                "user_or_group": user_or_group,
+            },
+        )
+        await cache.invalidate_func(
+            self.get_owned,
+            {
+                "artifact_type": namespace,
+                "name": name,
+                "user_or_group": user_or_group,
+            },
+        )
+        # Now rebuild the cache to ease the work on the next incoming message
+        await self.get_owners(namespace, project_name, user_or_group)
+        await self.get_owned(namespace, name, user_or_group)
+        tracked_cache = TrackedCache()
+        await tracked_cache.invalidate()
+        # We can't rebuild the tracked cache here because we don't have access to the DB and
+        # Requester objects.
```

Comment:
I'm really unsure about the entire concept of "refetch stuff we just invalidated", also about the implementation here:

- We don't know if a message related to the respective project and user/group comes in before the cache entry expires. If so, we cause load elsewhere unnecessarily. Wouldn't this be covered by priming the tracked cache anyway? I can't get rid of the feeling that this might be simpler if we cached closer to the backend service in question. E.g. (see my draft PR #717 😉), the Pagure proxy knows which of its methods deal with say projects/artifacts and can more “surgically” delete cached entries that would yield stale results after a certain change a message is about. Which, if we did it right, would obviate having to throw away and rebuild the whole tracked cache, I think.
- The `get_owners()` and `get_owned()` coroutines are awaited, but serially – we could make them tasks and `asyncio.gather()` them if we think Pagure can handle the load 😅.
- Invalidating the tracked cache without rebuilding it, in conjunction with the order of things in `Consumer.refresh_cache_if_needed()`, would fill the tracked cache and then invalidate it again (actually, invalidate it, the rebuild it, then invalidate it again, if I'm not mistaken).

Reply:
Interesting pointes!

- I'm not sure I understand this one. I'm fine with not rebuilding right after invalidating, because it'll only speed up another non-interactive process, so there's no much point. With regards to the tracked cache, I don't think we can remove parts of it: if I watch the kernel package and you watch it too, when I stopped watching it it should remain in the tracked set.

- Sure, hopefully that won't break Pagure :-)
- If I understand your comment correctly, I don't think that can happen: the messages that `TrackedCache.invalidate_on_message()` reacts to are not the messages that the distgit or fasjson services react to. Or did you mean something else?